File: Documents/gestion-stock/cpp-version/src/repository/ProduitRepository.cpp
#include "../../include/repository/ProduitRepository.hpp"
#include <algorithm>

void ProduitRepository::save(const Produit& produit) {
    produits[produit.getId()] = produit;
    produitsParCategorie[produit.getCategorieId()].push_back(produit.getId());
}

std::optional<Produit> ProduitRepository::findById(const std::string& id) const {
    auto it = produits.find(id);
    if (it != produits.end()) {
        return it->second;
    }
    return std::nullopt;
}

std::vector<Produit> ProduitRepository::findAll() const {
    std::vector<Produit> result;
    result.reserve(produits.size());
    for (const auto& pair : produits) {
        result.push_back(pair.second);
    }
    return result;
}

std::vector<Produit> ProduitRepository::findByCategorie(const std::string& categorieId) const {
    std::vector<Produit> result;
    auto it = produitsParCategorie.find(categorieId);
    if (it != produitsParCategorie.end()) {
        for (const auto& id : it->second) {
            auto produit = findById(id);
            if (produit.has_value()) {
                result.push_back(produit.value());
            }
        }
    }
    return result;
}

bool ProduitRepository::deleteById(const std::string& id) {
    auto produit = findById(id);
    if (!produit.has_value()) {
        return false;
    }
    
    // Supprimer de la catégorie
    auto& ids = produitsParCategorie[produit->getCategorieId()];
    ids.erase(std::remove(ids.begin(), ids.end(), id), ids.end());
    
    // Supprimer du map principal
    produits.erase(id);
    return true;
}

bool ProduitRepository::exists(const std::string& id) const {
    return produits.find(id) != produits.end();
}

std::vector<Produit> ProduitRepository::findEnStockInsuffisant() const {
    std::vector<Produit> result;
    for (const auto& pair : produits) {
        if (pair.second.estEnStockInsuffisant()) {
            result.push_back(pair.second);
        }
    }
    return result;
}

void ProduitRepository::clear() {
    produits.clear();
    produitsParCategorie.clear();
}

size_t ProduitRepository::size() const {
    return produits.size();
}

File: Documents/gestion-stock/web-version/src/repositories/ProduitRepository.js
class ProduitRepository {
    constructor() {
        this.produits = new Map();
        this.produitsParCategorie = new Map();
    }

    save(produit) {
        this.produits.set(produit.id, produit);
        if (!this.produitsParCategorie.has(produit.categorieId)) {
            this.produitsParCategorie.set(produit.categorieId, new Set());
        }
        this.produitsParCategorie.get(produit.categorieId).add(produit.id);
    }

    findById(id) {
        return this.produits.get(id) || null;
    }

    findAll() {
        return Array.from(this.produits.values());
    }

    findByCategorie(categorieId) {
        const ids = this.produitsParCategorie.get(categorieId) || new Set();
        return Array.from(ids).map(id => this.produits.get(id)).filter(Boolean);
    }

    deleteById(id) {
        const produit = this.produits.get(id);
        if (!produit) return false;
        
        const ids = this.produitsParCategorie.get(produit.categorieId);
        if (ids) ids.delete(id);
        
        this.produits.delete(id);
        return true;
    }

    exists(id) {
        return this.produits.has(id);
    }

    findEnStockInsuffisant() {
        return Array.from(this.produits.values()).filter(p => p.estEnStockInsuffisant());
    }

    clear() {
        this.produits.clear();
        this.produitsParCategorie.clear();
    }

    size() {
        return this.produits.size;
    }
}

module.exports = ProduitRepository;